src/sagas/wallet.js

```diff
@@ -0,0 +1,550 @@
+
+  // The way the redux-saga fork model works is that if a saga has `forked`
+  // another saga (using the `fork` effect), it will remain active until all
+  // the forks are terminated. You can read more details at
+  // https://redux-saga.js.org/docs/advanced/ForkModel
+  // So, if a new START_WALLET_REQUESTED action is dispatched, we need to cleanup
+  // all attached forks (that will cause the event listeners to be cleaned).
+  while (true) {
+    yield take('START_WALLET_REQUESTED');
+    yield cancel([
+      walletListenerThread,
+      walletReadyThread,
+      featureFlagsThread
+    ]);
+  }
+}
+
+/**
+ * This saga will load both HTR and DEFAULT_TOKEN (if they are different)
+ * and dispatch actions to asynchronously load all registered tokens
+ */
+export function* loadTokens() {
+  const customTokenUid = DEFAULT_TOKEN.uid;
+  const htrUid = hathorLibConstants.HATHOR_TOKEN_CONFIG.uid;
+
+  // Download hathor token balance and history:
+  yield put(tokenFetchBalanceRequested(hathorLibConstants.HATHOR_TOKEN_CONFIG.uid));
+  const { htrBalanceError } = yield race({
+    success: take(specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_SUCCESS, {
+      tokenId: htrUid,
+    })),
+    htrBalanceError: take(specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_FAILED, {
+      tokenId: htrUid,
+    })),
+  });
+  // ...and history:
+  yield put(tokenFetchHistoryRequested(hathorLibConstants.HATHOR_TOKEN_CONFIG.uid));
+  const { htrHistoryError } = yield race({
+    success: take(specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_SUCCESS, {
+      tokenId: htrUid,
+    })),
+    htrHistoryError: take(specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_FAILED, {
+      tokenId: htrUid,
+    })),
+  });
+
+  if (htrBalanceError || htrHistoryError) {
+    throw new Error('Error loading HTR history or balance');
+  }
+
+  if (customTokenUid !== htrUid) {
+    // Download custom token balance:
+    yield put(tokenFetchBalanceRequested(customTokenUid));
+    const { customBalanceError } = yield race({
+      success: take(specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_SUCCESS, {
+        tokenId: customTokenUid,
+      })),
+      customBalanceError: take(specificTypeAndPayload(types.TOKEN_FETCH_BALANCE_FAILED, {
+        tokenId: htrUid,
+      })),
+    });
+    // ...and history:
+    yield put(tokenFetchHistoryRequested(customTokenUid));
+    const { customHistoryError } = yield race({
+      success: take(specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_SUCCESS, {
+        tokenId: customTokenUid,
+      })),
+      customHistoryError: take(specificTypeAndPayload(types.TOKEN_FETCH_HISTORY_FAILED, {
+        tokenId: customTokenUid,
+      })),
+    });
+
+    if (customBalanceError || customHistoryError) {
+      throw new Error('Error loading custom token history or balance');
+    }
+  }
+
+  const registeredTokens = tokensUtils
+    .getTokens()
+    .reduce((acc, token) => {
+      // remove htr since we will always download the HTR token
```

Comment:
Should we remove the default token as well?

In one hand, is good so we don't download the balance/history of this token again (or do we check the status: ready?), on the other hand we need to fetch the metadata of this token anyway, so we need to take care here

Reply:
We do check the `status: ready`, so we won't download it again

I think it is fine to just dispatch the action and let the saga ignore it